lib/src/migrators/namespace.dart

```diff
@@ -0,0 +1,245 @@
+      ImportCache importCache, Stylesheet stylesheet, Importer importer) {
+    Renamer<UseRule> renamer;
+    try {
+      renamer = Renamer(argResults['rename'].join('\n'), 'namespace', {
+        'namespace': (rule) => rule.namespace,
+        'url': (rule) => rule.url.toString()
+      });
+    } on FormatException catch (e) {
+      throw MigrationException(e.message);
+    }
+    var visitor = _NamespaceMigrationVisitor(
+        renamer,
+        UnreferencedFlag(argResults['unreferenced']),
+        argResults['force'] as bool,
+        importCache,
+        migrateDependencies);
+    var result = visitor.run(stylesheet, importer);
+    missingDependencies.addAll(visitor.missingDependencies);
+    return result;
+  }
+}
+
+class _NamespaceMigrationVisitor extends MigrationVisitor {
+  final Renamer<UseRule> renamer;
+  final UnreferencedFlag unreferencedFlag;
+  final bool forceRename;
+
+  Map<String, Set<FileSpan>> _spansByNamespace;
+  Set<String> _usedNamespaces;
+
+  _NamespaceMigrationVisitor(this.renamer, this.unreferencedFlag,
+      this.forceRename, ImportCache importCache, bool migrateDependencies)
+      : super(importCache, migrateDependencies);
+
+  @override
+  void visitStylesheet(Stylesheet node) {
+    var oldSpansByNamespace = _spansByNamespace;
+    var oldUsedNamespaces = _usedNamespaces;
+    _spansByNamespace = {};
+    _usedNamespaces = {};
+    super.visitStylesheet(node);
+    _spansByNamespace = oldSpansByNamespace;
+    _usedNamespaces = oldUsedNamespaces;
+  }
+
+  @override
+  void beforePatch(Stylesheet node) {
+    // Pass each `@use` rule through
+    var newNamespaces = <String, Set<UseRule>>{};
+    for (var rule in node.children.whereType<UseRule>()) {
+      if (rule.namespace == null) continue;
+      newNamespaces
+          .putIfAbsent(renamer.rename(rule) ?? rule.namespace, () => {})
+          .add(rule);
+    }
+    var unreferencedRules = <UseRule>{};
+    for (var entry in newNamespaces.entries) {
+      var newNamespace = entry.key;
+      var rules = entry.value;
+      if (rules.length == 1) {
+        if (unreferencedFlag == UnreferencedFlag.all &&
+            !_spansByNamespace.containsKey(rules.first.namespace)) {
+          unreferencedRules.add(rules.first);
+        } else {
+          _patchNamespace(rules.first, newNamespace);
+        }
+        continue;
+      }
+      if (unreferencedFlag != UnreferencedFlag.none) {
+        for (var rule in rules.toSet()) {
+          if (!_spansByNamespace.containsKey(rule.namespace)) {
+            unreferencedRules.add(rule);
+            rules.remove(rule);
+          }
+        }
+      }
+      if (rules.length == 1) {
+        _patchNamespace(rules.first, newNamespace);
+        continue;
+      }
+      if (!forceRename) {
+        throw MigrationException(
+            'Rename failed. ${rules.length} rules would use namespace '
+            '"$newNamespace".\n'
+            'Run with --force to rename with numerical suffixes.');
```

Comment:
This should include the spans of the rules in question. You can hook into `MultiSpanSassException` to make that easy.